### PR TITLE
[Painless] Partially fixes def boxed types casting

### PR DESCRIPTION
**Pull request by jdconrad**
An issue related to implicitly casting to appropriate return types was revealed (https://github.com/elastic/elasticsearch/issues/35351).  According to the Painless spec (https://www.elastic.co/guide/en/elasticsearch/painless/current/painless-casting.html), def values are allowed to implicitly cast from one boxed type to another boxed type such as `Integer` to `Double`.  Currently, this not working according to spec.  This partially fixes the issue by inserting appropriate methods at compile-time to execute the casts at run-time.  The other part of the issue will be a different PR as it needs to fix an issue related to boxing for method parameters at run-time which is trickier due to run-time code path.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java [hunk 1]
```diff
@@ -706,62 +706,222 @@ public static double DefTodoubleImplicit(final Object value) {
...
+            return (double)(float)value;
+        } else {
+            return (Double)value;
+        }
+    }
+
+    public static Byte defToByteExplicit(final Object value) {
+        if (value == null) {
+            return null;
+        } else if (value instanceof Character) {
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java]: I thought byte -> char and char -> byte had to be explicit (in java). Does painless differ here?

**Review by rjernst — COMMENTED** [review]
Looks fine, just one question.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java]: Good catch, you are correct.  This is a bug that I'll fix now.

Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java]: Note this is the explicit version, though.  I corrected the implicit versions.

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by jdconrad** [comment]
@rjernst Thanks for the review!  Will commit as soon as CI passes.

**Comment by jdconrad** [comment]
I'm closing this in favor of the quick/easier fix for now (https://github.com/elastic/elasticsearch/pull/35653).  The casting model in Painless needs to be revisited in a separate follow up as some of the specced out casts do not work as expected, but the fixes need to be given more thought so more problems are not created.

**Pull request closed without merging**